### PR TITLE
Tracking for individual/ evaluated individual

### PR DESCRIPTION
**Pull request by ZhangMan1126**
The branch is for enabling tracking individual/ evaluated individual.

### CODE REVIEW

File: core/src/main/kotlin/org/evomaster/core/EMConfig.kt [hunk 1]
```diff
@@ -515,4 +519,14 @@ class EMConfig {
     @Experimental
     @Cfg("Enable EvoMaster to generate, use, and attach complete objects to REST calls, rather than just the needed fields/values")
     var enableCompleteObjects = false
+
+    @Experimental
+    @Cfg("Whether to track individual")
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/EMConfig.kt]: can you provide bit more description? eg. here I assume you are tracking the history of modifications of the individuals during the search

File: core/src/main/kotlin/org/evomaster/core/EMConfig.kt [hunk 2]
```diff
@@ -515,4 +519,14 @@ class EMConfig {
...
+    var enableTrackIndividual = false
+
+    @Experimental
+    @Cfg("Whether to track evaluated individual. " +
+            "Note that we recommend that set enableTrackIndividual false when enableTrackEvaluatedIndividual is true since information of individual is part of evalauted individual")
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/EMConfig.kt]: "recommend"? or enforced? don't you throw an exception if both are true?

File: core/src/main/kotlin/org/evomaster/core/problem/rest/RestIndividual.kt [hunk 3]
```diff
@@ -7,20 +7,45 @@ import org.evomaster.core.search.Action
...
+
+
+class RestIndividual : Individual {
+
+    val actions: MutableList<RestAction>
+    val sampleType: SampleType
+    val dbInitialization: MutableList<DbAction>
+    val usedObjects: UsedObjects
+
+    constructor(
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/rest/RestIndividual.kt]: why using a "constructor"? you know that you can have default parameters in the inputs?

File: core/src/main/kotlin/org/evomaster/core/search/EvaluatedIndividual.kt [hunk 4]
```diff
@@ -44,4 +57,55 @@ where T : Individual {
...
+                 */
+                getTrack()?:return EvaluatedIndividual(
+                        fitness.copy(),
+                        individual.copy(withTrack) as T,
+                        results.map(ActionResult::copy),
+                        trackOperator
+                )
+
+                return forceCopyWithTrack()
+
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/EvaluatedIndividual.kt]: unnecessary empty line?

File: core/src/main/kotlin/org/evomaster/core/search/Individual.kt [hunk 5]
```diff
@@ -2,21 +2,32 @@ package org.evomaster.core.search
...
+ * @property track is a list of Individual, indicating its evolution
  */
-abstract class Individual {
+abstract class Individual : TraceableElement{
+
+    constructor(trackOperator: TrackOperator? = null, traces : MutableList<out Individual>? = null) : super(trackOperator, traces)
 
     /**
-     * Make a deep copy of this individual
+     * Make a deep next of this individual
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/Individual.kt]: when you refactored the variable name from "copy" to "next", didn't you by mistake also did in the comments without double-checking it?

File: core/src/main/kotlin/org/evomaster/core/search/service/Archive.kt [hunk 6]
```diff
@@ -97,7 +101,7 @@ class Archive<T> where T : Individual {
     fun isEmpty() = populations.isEmpty()
 
     /**
-     * Get a copy of an individual in the archive.
+     * Get a next of an individual in the archive.
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/Archive.kt]: see problem above

File: core/src/main/kotlin/org/evomaster/core/search/service/Archive.kt [hunk 7]
```diff
@@ -359,14 +363,15 @@ class Archive<T> where T : Individual {
 
             if (better || equivalent) {
                 /*
-                    replace worst element, if copy is not worse than it (but not necessarily better).
+                    replace worst element, if next is not worse than it (but not necessarily better).
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/Archive.kt]: see previous problem

File: core/src/main/kotlin/org/evomaster/core/search/service/mutator/Mutator.kt [hunk 8]
```diff
@@ -30,7 +31,7 @@ abstract class Mutator<T> where T : Individual {
     protected lateinit var config: EMConfig
 
     /**
-     * @return a mutated copy
+     * @return a mutated next
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/mutator/Mutator.kt]: see previous problem

File: core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt [hunk 9]
```diff
@@ -51,7 +51,7 @@ class StandardMutator<T> : Mutator<T>() where T : Individual {
 
         var mutated = false
 
-        while (!mutated) { //no point in returning a copy that is not mutated
+        while (!mutated) { //no point in returning a next that is not mutated
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt]: see previous problem

File: core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt [hunk 10]
```diff
@@ -181,7 +183,7 @@ class StandardMutator<T> : Mutator<T>() where T : Individual {
 
     private fun handleDoubleGene(gene: DoubleGene) {
         //TODO min/max for Double
-
+        val d = gene.value
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt]: why this variable?

File: core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt [hunk 11]
```diff
@@ -218,6 +219,7 @@ class StandardMutator<T> : Mutator<T>() where T : Individual {
 
 
     private fun handleIntegerGene(gene: IntegerGene) {
+        val i = gene.value
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt]: why this variable?

File: core/src/test/kotlin/org/evomaster/core/search/algorithms/MioAlgorithmOnTrackOneMaxTest.kt [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    @Test
+    fun testIndividualWithTrack(){
+
+        config.enableTrackIndividual = true
+        config.enableTrackEvaluatedIndividual = false
+
+        val solution = mio.search()
+
+        solution.individuals.forEach { s->
+            Assertions.assertNull(s.getTrack())
```
Comment (arcuri82) [inline core/src/test/kotlin/org/evomaster/core/search/algorithms/MioAlgorithmOnTrackOneMaxTest.kt]: you do not need to repeat "Assertions." in each call. You can just do a static import of those static methods

File: core/src/main/kotlin/org/evomaster/core/EMConfig.kt (hunk 2 above)
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/EMConfig.kt]: modified.

File: core/src/main/kotlin/org/evomaster/core/problem/rest/RestIndividual.kt (hunk 3 above)
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/problem/rest/RestIndividual.kt]: modified. thanks! but somehow, the class cannot be created with default values in java. So I modified a bit in a test "DbDirectIntWithSqlEMTest" of e2e-test as below         

RestIndividual withSQL = new RestIndividual(ind.getActions(), ind.getSampleType(), insertions, ind.getUsedObjects(), null, null);

File: core/src/main/kotlin/org/evomaster/core/search/EvaluatedIndividual.kt (hunk 4 above)
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/search/EvaluatedIndividual.kt]: modified.

File: core/src/main/kotlin/org/evomaster/core/search/Individual.kt (hunk 5 above)
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/search/Individual.kt]: yes. it is a mistake.

File: core/src/main/kotlin/org/evomaster/core/search/service/mutator/Mutator.kt (hunk 8 above)
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/search/service/mutator/Mutator.kt]: modified

File: core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt (hunk 9 above)
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt]: modified

File: core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt (hunk 10 above)
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt]: thanks. it is a mistake. when i implemented archive-based mutation, i edit a bit. but i forgot to change it.

File: core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt (hunk 11 above)
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/search/service/mutator/StandardMutator.kt]: modified

File: core/src/test/kotlin/org/evomaster/core/search/algorithms/MioAlgorithmOnTrackOneMaxTest.kt (hunk 12 above)
Comment (ZhangMan1126) [reply core/src/test/kotlin/org/evomaster/core/search/algorithms/MioAlgorithmOnTrackOneMaxTest.kt]: thanks. modified

**Pull request merged**